### PR TITLE
drivers: serial: stm32: use generic LL headers 

### LINKED ISSUE #28822: Improve STM32 LL HAL usage
## Introduction

This RFC aims to improve the usage of STM32 LL HAL in both Zephyr drivers and for out-of-tree users.

### Problem description

STM32 drivers using LL API need to include `stm32xxx_ll_yyy.h` headers in order to make use of the API. Nowadays this is done via the `soc.h` file, e.g. https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/st_stm32/stm32f4/soc.h. Inclusion of LL headers is dependent on active configuration, e.g.

```c
...
#ifdef CONFIG_ADC_STM32
#include <stm32f4xx_ll_adc.h>
#endif

#ifdef CONFIG_DMA_STM32
#include <stm32f4xx_ll_dma.h>
#endif
...
```

This approach has some problems:
1. Every time a new driver or SoC is added `soc.h` needs to be adjusted accordingly (SoC headers depend on driver configuration)
2. When a driver includes `soc.h`, it will include all LL headers that are active. For example, if ADC and DMA are active, both will be included. STM32 LL API is mainly built upon `static inline` functions, so when we include all these headers we are *pasting* all of them in a driver that will, in most cases, just need one of the LL headers.
3. Out of tree applications using LL API can't rely on changing `soc.h`, so in case the application is designed for multiple series, it needs to include LL API based on the active SoC, e.g.

```c
#ifdef CONFIG_SOC_SERIES_STM32F3XX
#include <stm32f3xx_ll_dma.h>
#elif CONFIG_SOC_SERIES_STM32F4XX
#include <stm32f4xx_ll_dma.h>
#endif
```

Note that the non-LL HAL is not affected, as it already provides a header gathering all APIs. It can remain part of `soc.h`.

### Proposed change

My proposal is to provide SoC generic LL HAL headers, for example:

```c
/* stm32_ll_tim.h */
#include <autoconf.h>

#ifdef CONFIG_SOC_SERIES_STM32F0XX
#include <stm32f0xx_ll_tim.h>
...
#elif CONFIG_SOC_SERIES_STM32F4XX
#include <stm32f4xx_ll_tim.h>
...
#endif
```

and then the driver would just include:

```c
#include <stm32_ll_tim.h>
```

The advantages of this approach:
1. SoC dependency on drivers' configuration is removed
2. Only the necessary stuff gets included
3. Out of tree application will also have access to generic LL headers, making it easier to create multi-family applications.

### Proposed change (Detailed)

The first step would be to have the generic LL headers. Their generation can be scripted and included in the `stm32_hal` repository. The second step would be to take a reference driver, remove `soc.h` entries and include LL API directly in the driver. It should be a small change and can be done gradually to all other drivers. 

### Dependencies

To be discussed.

### Concerns and Unresolved Questions

To be discussed.

## Alternatives

N/A

### PR DESCRIPTION
**Pull request by gmarull**
This PR is a proposal to improve LL HAL API usage by introducing generic LL headers. The full details of the proposal can be found in #28822. Companion PR: https://github.com/zephyrproject-rtos/hal_stm32/pull/68

Resolves #28822

### CODE REVIEW

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -28,6 +28,14 @@
 #include <drivers/clock_control/stm32_clock_control.h>
 #include "uart_stm32.h"
 
+#include <stm32_ll_usart.h>
+#include <stm32_ll_lpuart.h>
```
Comment (lowlander) [inline drivers/serial/uart_stm32.c:32]: Several `soc.h` don't include `stm32*_ll_lpuart.h` header files at the moment, is this handled with dummy/empty include files in the `stm32_hal` repo ?

Comment (gmarull) [reply drivers/serial/uart_stm32.c:32]: @lowlander Yes. Generated headers only contain entries for families that have the API.

https://github.com/gmarull/hal_stm32/blob/5b5cfa63fa9b9f5f648bc20e1500ab377af85af8/include/stm32_ll_lpuart.h

**Comment by martinjaeger** [comment]
That's a great contribution to simplify adding driver support for new MCU series. And also helps to clean up my out-of-tree drivers. Definitely +1 from my side once the PR is merged to hal_stm32.

**Comment by Nukersson** [comment]
`west.yml`looks to have conflicts. Can you please provide proper changes to it? Otherwise LGTM! And good work!

**Comment by gmarull** [comment]
@erwango @ABOSTM should be ready now.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@gmarull, would you mind rebasing ?

**Comment by gmarull** [comment]
@erwango done

**Comment by erwango** [comment]
@lowlander, @martinjaeger, @Nukersson, @lochej this PR requires some more positive reviews to get in.

**Comment by Nukersson** [comment]
> @lowlander, @martinjaeger, @Nukersson, @lochej this PR requires some more positive reviews to get in.

Thank you. I will be able to review it only at the weekend.

**Review by martinjaeger — APPROVED** [review]
Thanks @gmarull. Looks good!

Are you planning to submit a PR for the other peripherals aswell?

**Comment by erwango** [comment]
> Are you planning to submit a PR for the other peripherals aswell?

Maybe that's an effort that can be shared

**Comment by gmarull** [comment]
> > Are you planning to submit a PR for the other peripherals aswell?
> 
> Maybe that's an effort that can be shared

@erwango agree, contributions are welcome :-)

**Review by KozhinovAlexander — APPROVED** [review]

**Comment by martinjaeger** [comment]
> > > Are you planning to submit a PR for the other peripherals aswell?
> > 
> > 
> > Maybe that's an effort that can be shared
> 
> @erwango agree, contributions are welcome :-)

Sure, I'm happy to support. Just didn't want to work on a PR if you have something in the pipeline already. I can look into a few peripherals this afternoon (let's say at least CAN, ADC, DAC).

**Comment by gmarull** [comment]
@martinjaeger sounds good!

**Review by carlescufi — APPROVED** [review]

**Pull request merged**